File: src/intelligence/validation.ts
import Ajv2020, {
  type AnySchema,
  type ErrorObject,
  type ValidateFunction,
} from "ajv/dist/2020.js";
import addFormats from "ajv-formats";

import {
  CURRENT_INTELLIGENCE_SCHEMA_VERSION,
  INTELLIGENCE_CONTRACT_KINDS,
  INTELLIGENCE_CONTRACTS,
  type IntelligenceContractDocuments,
  type IntelligenceContractKind,
  type IntelligenceJsonObject,
  type MigratedIntelligenceDocument,
} from "./contracts";

export const MAX_INTELLIGENCE_VALIDATION_ISSUES = 64;
const MAX_ISSUE_PATH_LENGTH = 256;
const SUPPORTED_SCHEMA_MAJOR = 1;

export type IntelligenceValidationErrorCode =
  | "invalid_schema_version"
  | "unsupported_schema_version"
  | "schema_validation_failed";

export interface IntelligenceValidationIssue {
  readonly code: "schema_mismatch";
  readonly path: string;
  readonly message: string;
}

export interface IntelligenceValidationFailure<
  Kind extends IntelligenceContractKind,
> {
  readonly ok: false;
  readonly kind: Kind;
  readonly validated: false;
  readonly code: IntelligenceValidationErrorCode;
  readonly issues: readonly IntelligenceValidationIssue[];
}

export type IntelligenceMigrationResult<
  Kind extends IntelligenceContractKind,
> =
  | {
      readonly ok: true;
      readonly kind: Kind;
      readonly validated: false;
      readonly value: MigratedIntelligenceDocument;
    }
  | IntelligenceValidationFailure<Kind>;

export type IntelligenceValidationResult<
  Kind extends IntelligenceContractKind,
> =
  | {
      readonly ok: true;
      readonly kind: Kind;
      readonly validated: true;
      readonly value: IntelligenceContractDocuments[Kind];
    }
  | IntelligenceValidationFailure<Kind>;

type MinorMigration = (
  document: IntelligenceJsonObject,
) => IntelligenceJsonObject;

// Adding a supported minor revision requires an explicit, reviewable entry.
// The current-version identity migration is intentionally idempotent.
const MINOR_MIGRATION_REGISTRY: ReadonlyMap<string, MinorMigration> = new Map([
  [CURRENT_INTELLIGENCE_SCHEMA_VERSION, (document) => document],
]);

const ajv = new Ajv2020({
  allErrors: true,
  strict: true,
  validateFormats: true,
});
addFormats(ajv);

for (const kind of INTELLIGENCE_CONTRACT_KINDS) {
  const contract = INTELLIGENCE_CONTRACTS[kind];
  ajv.addSchema(contract.schema as AnySchema, contract.schemaId);
}

const validators = Object.fromEntries(
  INTELLIGENCE_CONTRACT_KINDS.map((kind) => {
    const validator = ajv.getSchema(INTELLIGENCE_CONTRACTS[kind].schemaId);
    if (validator === undefined) {
      throw new Error("intelligence_schema_unavailable");
    }
    return [kind, validator];
  }),
) as Readonly<Record<IntelligenceContractKind, ValidateFunction>>;

function isJsonObject(value: unknown): value is IntelligenceJsonObject {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function parseVersion(value: unknown): readonly [number, number, number] | null {
  if (typeof value !== "string" || value.length > 16) {
    return null;
  }
  const match = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$/.exec(value);
  if (match === null) {
    return null;
  }
  const parts = match.slice(1).map(Number);
  if (parts.some((part) => !Number.isSafeInteger(part))) {
    return null;
  }
  return [parts[0], parts[1], parts[2]];
}

function versionFailure<Kind extends IntelligenceContractKind>(
  kind: Kind,
  code: Extract<
    IntelligenceValidationErrorCode,
    "invalid_schema_version" | "unsupported_schema_version"
  >,
): IntelligenceValidationFailure<Kind> {
  return {
    ok: false,
    kind,
    validated: false,
    code,
    issues: [
      {
        code: "schema_mismatch",
        path: "/schemaVersion",
        message:
          code === "invalid_schema_version"
            ? "Schema version is missing or malformed."
            : "Schema version is not supported by this runtime.",
      },
    ],
  };
}

export function migrateIntelligenceContract<
  Kind extends IntelligenceContractKind,
>(kind: Kind, value: unknown): IntelligenceMigrationResult<Kind> {
  if (!isJsonObject(value)) {
    return versionFailure(kind, "invalid_schema_version");
  }
  const parsedVersion = parseVersion(value.schemaVersion);
  if (parsedVersion === null) {
    return versionFailure(kind, "invalid_schema_version");
  }
  if (parsedVersion[0] !== SUPPORTED_SCHEMA_MAJOR) {
    return versionFailure(kind, "unsupported_schema_version");
  }

  const migration =
    typeof value.schemaVersion === "string"
      ? MINOR_MIGRATION_REGISTRY.get(value.schemaVersion)
      : undefined;
  if (migration === undefined) {
    return versionFailure(kind, "unsupported_schema_version");
  }

  return {
    ok: true,
    kind,
    validated: false,
    value: migration(value) as MigratedIntelligenceDocument,
  };
}

function boundedText(value: string, maximumLength: number): string {
  let result = "";
  for (const character of value) {
    if (result.length + character.length > maximumLength) {
      break;
    }
    result += character;
  }
  return result;
}

function issueMessage(keyword: string): string {
  switch (keyword) {
    case "additionalProperties":
      return "Property is not allowed by the committed contract.";
    case "required":
      return "A required property is missing from the contract.";
    case "format":
      return "Value does not satisfy the required public format.";
    case "const":
      return "Value does not match the committed contract version.";
    default:
      return "Value does not satisfy the committed contract.";
  }
}

function schemaIssue(error: ErrorObject): IntelligenceValidationIssue {
  return {
    code: "schema_mismatch",
    path: boundedText(error.instancePath || "/", MAX_ISSUE_PATH_LENGTH),
    message: issueMessage(error.keyword),
  };
}

export function validateIntelligenceContract<
  Kind extends IntelligenceContractKind,
>(kind: Kind, value: unknown): IntelligenceValidationResult<Kind> {
  const migrated = migrateIntelligenceContract(kind, value);
  if (!migrated.ok) {
    return migrated;
  }

  const validator = validators[kind];
  if (validator(migrated.value)) {
    return {
      ok: true,
      kind,
      validated: true,
      value: migrated.value as IntelligenceContractDocuments[Kind],
    };
  }

  const issues = (validator.errors ?? [])
    .slice(0, MAX_INTELLIGENCE_VALIDATION_ISSUES)
    .map(schemaIssue);
  return {
    ok: false,
    kind,
    validated: false,
    code: "schema_validation_failed",
    issues,
  };
}

export function validateGameOperatingModel(
  value: unknown,
): IntelligenceValidationResult<"gameOperatingModel"> {
  return validateIntelligenceContract("gameOperatingModel", value);
}

File: src-tauri/src/intelligence/validation.rs
use std::{collections::HashMap, error::Error, fmt, sync::OnceLock};

use jsonschema::{Draft, Retrieve, Uri, Validator};
use serde_json::Value;

use super::contracts::{
    supported_schema_versions, ContractKind, CURRENT_SCHEMA_VERSION, INTELLIGENCE_SCHEMA_DIALECT,
    SUPPORTED_SCHEMA_MAJOR,
};

pub const MAX_VALIDATION_ISSUES: usize = 64;
const MAX_ISSUE_PATH_LENGTH: usize = 256;
const MAX_ISSUE_MESSAGE_LENGTH: usize = 160;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
enum ContractErrorCode {
    InvalidSchemaVersion,
    UnsupportedSchemaVersion,
    SchemaUnavailable,
    SchemaValidationFailed,
}

impl ContractErrorCode {
    const fn as_str(self) -> &'static str {
        match self {
            Self::InvalidSchemaVersion => "invalid_schema_version",
            Self::UnsupportedSchemaVersion => "unsupported_schema_version",
            Self::SchemaUnavailable => "schema_unavailable",
            Self::SchemaValidationFailed => "schema_validation_failed",
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ContractValidationIssue {
    path: String,
    message: &'static str,
}

impl ContractValidationIssue {
    pub fn path(&self) -> &str {
        &self.path
    }

    pub const fn message(&self) -> &'static str {
        self.message
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ContractValidationError {
    code: ContractErrorCode,
    issues: Vec<ContractValidationIssue>,
}

impl ContractValidationError {
    fn new(code: ContractErrorCode, issues: Vec<ContractValidationIssue>) -> Self {
        Self {
            code,
            issues: issues.into_iter().take(MAX_VALIDATION_ISSUES).collect(),
        }
    }

    fn version(code: ContractErrorCode) -> Self {
        let message = match code {
            ContractErrorCode::InvalidSchemaVersion => "Schema version is missing or malformed.",
            ContractErrorCode::UnsupportedSchemaVersion => {
                "Schema version is not supported by this runtime."
            }
            _ => "Schema version could not be validated.",
        };
        Self::new(
            code,
            vec![ContractValidationIssue {
                path: "/schemaVersion".to_owned(),
                message,
            }],
        )
    }

    fn unavailable() -> Self {
        Self::new(ContractErrorCode::SchemaUnavailable, Vec::new())
    }

    pub const fn code(&self) -> &'static str {
        self.code.as_str()
    }

    pub fn issues(&self) -> &[ContractValidationIssue] {
        &self.issues
    }
}

impl fmt::Display for ContractValidationError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            formatter,
            "{}: intelligence contract rejected ({} issue(s))",
            self.code(),
            self.issues.len()
        )
    }
}

impl Error for ContractValidationError {}

type MigrationFn = fn(&Value) -> Value;

struct MinorMigration {
    source_version: &'static str,
    target_version: &'static str,
    migrate: MigrationFn,
}

fn identity_migration(document: &Value) -> Value {
    document.clone()
}

// Every supported minor revision must be registered explicitly. The current
// identity entry is deliberate: applying it repeatedly has the same result.
const MINOR_MIGRATION_REGISTRY: &[MinorMigration] = &[MinorMigration {
    source_version: CURRENT_SCHEMA_VERSION,
    target_version: CURRENT_SCHEMA_VERSION,
    migrate: identity_migration,
}];

pub fn migrate_contract(
    _kind: ContractKind,
    document: &Value,
) -> Result<Value, ContractValidationError> {
    let version = document
        .as_object()
        .and_then(|object| object.get("schemaVersion"))
        .and_then(Value::as_str)
        .ok_or_else(|| ContractValidationError::version(ContractErrorCode::InvalidSchemaVersion))?;

    let (major, _, _) = parse_schema_version(version)
        .ok_or_else(|| ContractValidationError::version(ContractErrorCode::InvalidSchemaVersion))?;
    if major != SUPPORTED_SCHEMA_MAJOR {
        return Err(ContractValidationError::version(
            ContractErrorCode::UnsupportedSchemaVersion,
        ));
    }

    let migration = MINOR_MIGRATION_REGISTRY
        .iter()
        .find(|registration| {
            registration.source_version == version
                && supported_schema_versions().contains(&registration.target_version)
        })
        .ok_or_else(|| {
            ContractValidationError::version(ContractErrorCode::UnsupportedSchemaVersion)
        })?;
    Ok((migration.migrate)(document))
}

pub fn validate_contract(
    kind: ContractKind,
    document: &Value,
) -> Result<Value, ContractValidationError> {
    let migrated = migrate_contract(kind, document)?;
    let validator = validators()?
        .get(&kind)
        .ok_or_else(ContractValidationError::unavailable)?;

    let mut issues = validator
        .iter_errors(&migrated)
        .map(|error| ContractValidationIssue {
            path: bounded_text(error.instance_path().as_str(), MAX_ISSUE_PATH_LENGTH),
            message: bounded_static_message(
                "Value does not satisfy the committed intelligence contract.",
            ),
        })
        .take(MAX_VALIDATION_ISSUES)
        .collect::<Vec<_>>();
    issues.sort_by(|left, right| left.path.cmp(&right.path));
    issues.dedup();

    if issues.is_empty() {
        Ok(migrated)
    } else {
        Err(ContractValidationError::new(
            ContractErrorCode::SchemaValidationFailed,
            issues,
        ))
    }
}

pub fn validate_game_operating_model(document: &Value) -> Result<Value, ContractValidationError> {
    validate_contract(ContractKind::GameOperatingModel, document)
}

fn parse_schema_version(value: &str) -> Option<(u64, u64, u64)> {
    if value.is_empty() || value.len() > 16 {
        return None;
    }
    let mut components = value.split('.');
    let major = parse_version_component(components.next()?)?;
    let minor = parse_version_component(components.next()?)?;
    let patch = parse_version_component(components.next()?)?;
    if components.next().is_some() {
        return None;
    }
    Some((major, minor, patch))
}

fn parse_version_component(value: &str) -> Option<u64> {
    if value.is_empty()
        || !value.bytes().all(|byte| byte.is_ascii_digit())
        || (value.len() > 1 && value.starts_with('0'))
    {
        return None;
    }
    value.parse().ok()
}

fn bounded_text(value: &str, maximum_bytes: usize) -> String {
    if value.len() <= maximum_bytes {
        return if value.is_empty() {
            "/".to_owned()
        } else {
            value.to_owned()
        };
    }
    let boundary = value
        .char_indices()
        .map(|(index, _)| index)
        .take_while(|index| *index <= maximum_bytes)
        .last()
        .unwrap_or(0);
    value[..boundary].to_owned()
}

const fn bounded_static_message(message: &'static str) -> &'static str {
    assert!(message.len() <= MAX_ISSUE_MESSAGE_LENGTH);
    message
}

static VALIDATORS: OnceLock<Result<HashMap<ContractKind, Validator>, ()>> = OnceLock::new();

fn validators() -> Result<&'static HashMap<ContractKind, Validator>, ContractValidationError> {
    VALIDATORS
        .get_or_init(build_validators)
        .as_ref()
        .map_err(|_| ContractValidationError::unavailable())
}

fn build_validators() -> Result<HashMap<ContractKind, Validator>, ()> {
    let mut validators = HashMap::with_capacity(ContractKind::ALL.len());
    for kind in ContractKind::ALL {
        let schema: Value = serde_json::from_str(kind.schema_source()).map_err(|_| ())?;
        if schema.get("$schema").and_then(Value::as_str) != Some(INTELLIGENCE_SCHEMA_DIALECT)
            || schema.get("$id").and_then(Value::as_str) != Some(kind.schema_id())
        {
            return Err(());
        }
        let validator = jsonschema::options()
            .with_draft(Draft::Draft202012)
            .should_validate_formats(true)
            .with_retriever(CommittedSchemaRetriever)
            .build(&schema)
            .map_err(|_| ())?;
        validators.insert(kind, validator);
    }
    Ok(validators)
}

#[derive(Clone, Copy)]
struct CommittedSchemaRetriever;

impl Retrieve for CommittedSchemaRetriever {
    fn retrieve(&self, uri: &Uri<String>) -> Result<Value, Box<dyn Error + Send + Sync>> {
        let kind = ContractKind::ALL
            .into_iter()
            .find(|kind| kind.schema_id() == uri.as_str())
            .ok_or_else(|| Box::new(SchemaRetrievalDenied) as Box<dyn Error + Send + Sync>)?;
        serde_json::from_str(kind.schema_source())
            .map_err(|_| Box::new(SchemaRetrievalDenied) as Box<dyn Error + Send + Sync>)
    }
}

#[derive(Debug)]
struct SchemaRetrievalDenied;

impl fmt::Display for SchemaRetrievalDenied {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str("schema retrieval denied")
    }
}

impl Error for SchemaRetrievalDenied {}
